[tf.data] graduate OptimizationOptions from experimental to tf.data

kvignesh1420 opened the pull request:
This PR graduates `tf.data.experimental.OptimizationOptions` to `tf.data.OptimizationOptions`.

- [x] deprecate the old endpoint
- [x] update docstring of the class
- [x] Modify the `__getattr__` and `__setattr__` implementation of `dataset_ops.Options` for backward-compatibility
- [x] regenerate golden APIs.
- [x] add/modify relevant test cases for the new endpoint.
- [x] update RELEASE.md

TEST LOG
```
INFO: Elapsed time: 32.496s, Critical Path: 28.35s
INFO: 354 processes: 133 internal, 221 local.
INFO: Build completed successfully, 354 total actions
//tensorflow/python/data/kernel_tests:options_test                       PASSED in 4.3s

INFO: Build completed successfully, 354 total actions
```

cc: @jsimsa

kvignesh1420 committed:
[tf.data] deprecate experimental OptimizationOptions

kvignesh1420 committed:
modify tests in options_test with optimization attr

kvignesh1420 committed:
regenerate golden APIs

jsimsa commented:
@kvignesh1420 thank you for the PR. FYI, I am working on removing some unused experimental optimization options that I would prefer to merge before your CL (so that we avoid moving some options only to remove them soon after).

On a related note, the branch cut for TF 2.6 is scheduled for 6/24. I am hoping to get my changes in in the next couple of days (and then your PR in before the branch cut).

Also, please take a look at the changes I made to your ThreadingOptions PR (https://github.com/tensorflow/tensorflow/commit/120ae92ab198392c7092176fd60a5049e0b7ac5a). The original approach (of maintaining both `threading` and `experimental_threading`) broke internal tests and so I switched to only keeping `threading` and having `experimental_threading` being re-routed to `threading`.

jsimsa commented on RELEASE.md:
```diff
@@ -122,8 +122,6 @@
         `tf.data.Dataset.shapshot` and deprecating the experimental endpoint.
     *   Promoting `tf.data.experimental.take_while` API to
         `tf.data.Dataset.take_while` and deprecating the experimental endpoint.
-    *   Promoting `tf.data.experimental.ThreadingOptions` API to
```
undo

kvignesh1420 replied:
done

jsimsa commented on tensorflow/python/data/experimental/ops/optimization_options.py:
```diff
@@ -29,19 +30,19 @@ class _AutotuneAlgorithm(enum.Enum):
 
-
-@tf_export("data.experimental.OptimizationOptions")
+@deprecation.deprecated_endpoints("data.experimental.OptimizationOptions")
```
Now that the OptimizationOptions will not be an experimental attribute of Options, some of the OptimizationOptions attribute should have an experimental prefix (so that they are not subject to backwards compatibility).

This includes:
- `filter_fusion`
- `filter_with_random_uniform_fusion`
- `hoist_random_uniform`
- `map_and_filter_fusion`
- `map_fusion`
- `reorder_data_discarding_ops`
- `parallel_batch` (which should be renamed to `experimental_batch_copy_parallelization`)

Furthermore, the autotuning options should be moved to a separate `AutotuneOptions` options object, which contains the following attributes:
- `enabled`
- `experimental_autotune_buffers`
- `cpu_budget`
- `memory_budget`

and is added to the top-level Options object as `autotune` attribute.

kvignesh1420 committed:
[tf.data] add new class and attribute for AutotuneOptions

add AutotuneOptions to __init__.py

add autotune_options dep to dataset_ops target

regenerate golden APIs for AutotuneOptions

fix AutotuneOptions protobuf instance call

kvignesh1420 committed:
[tf.data] add autotune_options based options_tests

kvignesh1420 commented:
@jsimsa I have made the initial python level changes for the new `AutotuneOptions` class and `autotune_options` Options attribute. Will make the changes to the `.cc` files after your CL.

jsimsa commented on RELEASE.md:
```diff
@@ -142,6 +142,8 @@
     *   Removed experimental tf.data API for map vectorization:
         *   `tf.data.experimental.OptimizationOptions.map_vectorization`
         *   `tf.data.experimental.MapVectorizationOptions.*`
+    *   Promoting `tf.data.experimental.OptimizationOptions` API to
```
we should also document the introducing of `tf.data.AutotuneOptions`

kvignesh1420 replied:
addressed in https://github.com/tensorflow/tensorflow/pull/50291

jsimsa commented on tensorflow/python/data/experimental/ops/autotune_options.py:
```diff
@@ -0,0 +1,98 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""API for controlling performance auto-tuning in `tf.data` pipelines."""
```
replace "auto-tuning" / "auto-tune" with "autotuning" / "autotune" throughout this PR.

kvignesh1420 replied:
addressed in https://github.com/tensorflow/tensorflow/pull/50291

jsimsa commented:
could we separate the `AutotuneOptions` change into a separate PR (which we could merge ahead of this PR) that would also include the C++ changes? thanks

kvignesh1420 commented:
@jsimsa sure will create a new PR for AutotuneOptions change.

kvignesh1420 commented:
@jsimsa  created the PR https://github.com/tensorflow/tensorflow/pull/50291 for addressing `AutotuneOptions`

gbaned commented:
@kvignesh1420 Can you please resolve conflicts? Thanks!

gbaned commented:
@kvignesh1420  Any update on this PR? Please. Thanks!

kvignesh1420 commented:
@jsimsa as you have covered the autotune options promotion in https://github.com/tensorflow/tensorflow/commit/c64634eb47f6581a7cf0cb01cf30284ff480d330, shall I continue with this PR? Please let me know.

cc: @aaudiber

jsimsa commented:
I would prefer to handle options related changes myself. My suggestion is to close this PR and prioritize other work suggested by @aaudiber (e.g. migrating `sample_from_datasets` to core API so that `rejection_resampling` implementation can stop depending on experimental API).

kvignesh1420 commented:
@jsimsa okay, sure!
